File: app/src/main/java/com/relic/data/gateway/PostGatewayImpl.java
package com.relic.data.gateway;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.content.Context;
import android.os.AsyncTask;
import android.util.Log;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.VolleyError;
import com.relic.R;
import com.relic.data.ApplicationDB;
import com.relic.data.Request.RedditOauthRequest;
import com.relic.data.VolleyAccessor;

public class PostGatewayImpl implements  PostGateway {
  private final String ENDPOINT = "https://oauth.reddit.com/";
  public static String TAG = "POST_GATEWAY";
  private String authToken;

  private ApplicationDB appDb;

  RequestQueue requestQueue;


  public PostGatewayImpl(Context context) {
    appDb = ApplicationDB.getDatabase(context);
    // Get the key values needed to get the actual authtoken from shared preferences
    String authKey = context.getString(R.string.AUTH_PREF);
    String tokenKey = context.getString(R.string.TOKEN_KEY);

    // retrieve the authtoken for use
    authToken = context.getSharedPreferences(authKey, Context.MODE_PRIVATE)
        .getString(tokenKey, "DEFAULT");

    requestQueue = VolleyAccessor.getInstance(context).getRequestQueue();
  }

  @Override
  public LiveData<Boolean> voteOnPost(String fullname, int voteStatus) {
    // generate the voting endpoint
    String ending = ENDPOINT + "api/vote?id=" + fullname;
    ending += "&dir=" + voteStatus;

    MutableLiveData<Boolean> success = new MutableLiveData<>();

    requestQueue.add(new RedditOauthRequest(Request.Method.POST, ending,
        response -> {
          Log.d(TAG, "Success voting on post : " + fullname + " to " + voteStatus);
          success.setValue(true);

          // update the local model appropriately
          new UpdateVoteStatus().execute(appDb, fullname, voteStatus);
        },
        (VolleyError error) -> {
          Log.d(TAG, "Sorry, there was an error voting on the post " + fullname + " to " + voteStatus);
          success.setValue(false);
        }, authToken));

    return success;
  }


  @Override
  public LiveData<Boolean> savePost(String fullname, boolean saved) {
    MutableLiveData<Boolean> success = new MutableLiveData<>();

    // generate the voting endpoint
    String ending = ENDPOINT + "api/save?id=" + fullname;

    requestQueue.add(new RedditOauthRequest(Request.Method.POST, ending,
        response -> {
          Log.d(TAG, "Success post saved status for " + fullname + " to " + saved);
          success.setValue(true);

          // update the local model appropriately
          new UpdateSaveStatus().execute(appDb, fullname, saved);
        },
        (VolleyError error) -> {
          Log.d(TAG, "Sorry, there was an error saving the post " + fullname);
          success.setValue(false);
        }, authToken));

    return success;
  }


  @Override
  public LiveData<Boolean> comment(String fullname, String comment) {
    MutableLiveData<Boolean> success = new MutableLiveData<>();

    return null;
  }

  @Override
  public LiveData<Boolean> gildPost(String fullname, boolean gild) {
    MutableLiveData<Boolean> success = new MutableLiveData<>();

    return null;
  }

  @Override
  public LiveData<Boolean> reportPosts(String fullname, boolean report) {
    MutableLiveData<Boolean> success = new MutableLiveData<>();

    return null;
  }

  @Override
  public LiveData<Boolean> visitPost(String postFullname) {
    Log.d(TAG, "Setting " + postFullname + "to visited");
    new UpdateVisitStatus().execute(appDb, postFullname);
    return null;
  }

  private static class UpdateVisitStatus extends AsyncTask<Object, Integer, Integer> {
    @Override
    protected Integer doInBackground(Object... objects) {
      ApplicationDB appDb = (ApplicationDB) objects[0];
      appDb.getPostDao().updateVisited((String) objects[1]);
      return null;
    }
  }

  private static class UpdateVoteStatus extends AsyncTask<Object, Integer, Integer> {
    @Override
    protected Integer doInBackground(Object... objects) {
      ApplicationDB appDb = (ApplicationDB) objects[0];
      appDb.getPostDao().updateVote((String) objects[1], (int) objects[2]);
      return null;
    }
  }

  private static class UpdateSaveStatus extends  AsyncTask<Object, Integer, Integer> {
    @Override
    protected Integer doInBackground(Object... objects) {
      ApplicationDB appDb = (ApplicationDB) objects[0];
      String postFullname = (String) objects[1];
      boolean saveStatus = (boolean) objects[2];
      appDb.getPostDao().updateSave(postFullname, saveStatus);
      return null;
    }
  }

}

File: app/src/main/java/com/relic/data/gateway/PostGatewayImpl.kt
package com.relic.data.gateway

import android.util.Log

import com.relic.persistence.ApplicationDB
import com.relic.data.DomainTransfer
import com.relic.persistence.entities.PostVisitRelation
import com.relic.data.repository.RepoConstants
import com.relic.network.NetworkRequestManager
import com.relic.network.request.RelicOAuthRequest
import com.squareup.moshi.Moshi
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import timber.log.Timber
import javax.inject.Inject

class PostGatewayImpl @Inject constructor(
    private val appDB : ApplicationDB,
    private val requestManager: NetworkRequestManager,
    private val moshi : Moshi
) : PostGateway {
    var TAG = "POST_GATEWAY"

    override suspend fun voteOnPost(fullname: String, voteStatus: Int) {
        // generate the voting endpoint
        val ending = RepoConstants.ENDPOINT + "api/vote?id=" + fullname + "&dir=$voteStatus"
        try {
            requestManager.processRequest(RelicOAuthRequest.POST, ending)
            Log.d(TAG, "Success voting on post : $fullname to $voteStatus")

            // update the local model appropriately
            withContext(Dispatchers.IO) {
                appDB.postDao.updateVote(fullname, voteStatus)
            }
        } catch (e : Exception) {
            throw DomainTransfer.handleException("vote on post", e) ?: e
        }
    }

    override suspend fun savePost(fullname: String, save: Boolean) {
        // generate the voting endpoint
        val saveString = if (save) "save" else "unsave"
        val ending = RepoConstants.ENDPOINT + "api/" + saveString + "?id=" + fullname
        try {
            requestManager.processRequest(RelicOAuthRequest.POST, ending)
            Log.d(TAG, "Success post saved status for $fullname to $save")

            // update the local model appropriately
            withContext(Dispatchers.IO) {
                appDB.postDao.updateSave(fullname, save)
            }
        } catch (e : Exception) {
            throw DomainTransfer.handleException("save post", e) ?: e
        }
    }


    override suspend fun comment(fullname: String, comment: String) {
        // TODO
    }

    override suspend fun gildPost(fullname: String, gild: Boolean) {
        // TODO
    }

    override suspend fun reportPosts(fullname: String, report: Boolean) {
        // TODO
    }

    override suspend fun visitPost(postFullname: String){
        Timber.d("Setting {postFullname} to visited")
        withContext(Dispatchers.IO) {
            appDB.postVisitedDao.insertVisited(PostVisitRelation(postFullname))
        }
    }
}
